Add leak tests for NodeIterator and TreeWalker

rniwa opened the pull request:
#### b17ccd4f3712b8d42d1113ec3f26423005b28806
<pre>
Add leak tests for NodeIterator and TreeWalker
<a href="https://bugs.webkit.org/show_bug.cgi?id=119718">https://bugs.webkit.org/show_bug.cgi?id=119718</a>

Reviewed by Darin Adler, Geoffrey Garen and Chris Dumez.

Add the leak tests based on <a href="https://chromium.googlesource.com/chromium/blink/+/65db79784ecd4a7e744d10a26befe5159638a56c">https://chromium.googlesource.com/chromium/blink/+/65db79784ecd4a7e744d10a26befe5159638a56c</a>
These tests have been modified to account for the fact WebKit / JSC uses a conservative GC as opposed to V8&apos;s precise GC.

* LayoutTests/fast/dom/NodeIterator/NodeIterator-dont-overcollect-expected.txt: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-dont-overcollect.html: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-leak-document-expected.txt: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-leak-document.html: Added.
* LayoutTests/fast/dom/TreeWalker/TreeWalker-leak-document-expected.txt: Added.
* LayoutTests/fast/dom/TreeWalker/TreeWalker-leak-document.html: Added.
* LayoutTests/fast/dom/resources/leak-check.js: Added.
(getCounterValues):
(loadSourceIntoIframe):
(async doLeakTest):
(htmlToUrl):
(grabScriptText):

Canonical link: <a href="https://commits.webkit.org/254711@main">https://commits.webkit.org/254711@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8db4a1ca8184fa0a9b1cbf386b780294bf1ac37f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89832 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99148 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155970 "Found 1 new test failure: fast/workers/worker-crash-with-invalid-location.html (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32870 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28310 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82203 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93498 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95479 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26118 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76651 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26064 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80994 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80838 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69071 "analyze-api-test-results") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30620 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14921 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30361 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15862 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33818 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38813 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1413 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34946 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/fast/dom/NodeIterator/NodeIterator-leak-document-expected.txt:
```diff
@@ -0,0 +1,5 @@
+PASS endingDocumentCount - startingDocumentCount < testingCount * 0.8 is true
```
Would be nice for have a `description();` in all these tests.

rniwa replied:
Good point.

darinadler commented on LayoutTests/fast/dom/NodeIterator/NodeIterator-dont-overcollect.html:
```diff
@@ -0,0 +1,24 @@
+<script src="../../../resources/js-test-pre.js"></script>
+<script src="../../../resources/gc.js"></script>
+<script>
+    description("This test asserts that NodeItrator's callback is kept alive while NodeIterator is alive even if all JS references go away.")
```
There is a typo here in one case: `NodeItrator`.

rniwa replied:
Fixed.

geoffreygaren commented on LayoutTests/fast/dom/resources/leak-check.js:
```diff
@@ -0,0 +1,47 @@
+        await loadSourceIntoIframe('about:blank');
+    }
+    endingDocumentCount = internals.numberOfLiveDocuments();
+    shouldBeTrue('endingDocumentCount - startingDocumentCount < testingCount * 0.8');
```
This is probably much more liberal than you need to be for conservative GC. But I guess it's OK for a leaks test because it does still prove the number is not 100%.

geoffreygaren reviewed (approved):
r=me

rniwa committed:
Add leak tests for NodeIterator and TreeWalker
https://bugs.webkit.org/show_bug.cgi?id=119718

Reviewed by Darin Adler, Geoffrey Garen and Chris Dumez.

Add the leak tests based on https://chromium.googlesource.com/chromium/blink/+/65db79784ecd4a7e744d10a26befe5159638a56c
These tests have been modified to account for the fact WebKit / JSC uses a conservative GC as opposed to V8's precise GC.

* LayoutTests/fast/dom/NodeIterator/NodeIterator-dont-overcollect-expected.txt: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-dont-overcollect.html: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-leak-document-expected.txt: Added.
* LayoutTests/fast/dom/NodeIterator/NodeIterator-leak-document.html: Added.
* LayoutTests/fast/dom/TreeWalker/TreeWalker-leak-document-expected.txt: Added.
* LayoutTests/fast/dom/TreeWalker/TreeWalker-leak-document.html: Added.
* LayoutTests/fast/dom/resources/leak-check.js: Added.
(getCounterValues):
(loadSourceIntoIframe):
(async doLeakTest):
(htmlToUrl):
(grabScriptText):

Canonical link: https://commits.webkit.org/254711@main
